packages/headless/src/features/search-parameters/search-parameter-serializer.ssr.test.ts

```diff
@@ -0,0 +1,137 @@
+
+  describe('when the search parameter value is a string', () => {
+    it('should convert URLSearchParams to SearchParameters', () => {
+      const urlSearchParams = new URLSearchParams();
+      urlSearchParams.append('q', 'query');
+      urlSearchParams.append('f-color', 'red');
+      urlSearchParams.append('f-shape', 'square');
+
+      const result = toSearchParameters(urlSearchParams);
+
+      expect(result).toEqual({
+        q: 'query',
+        f: {color: ['red'], shape: ['square']},
+      });
+    });
+
+    it('should convert Record<string, SearchParamValue> to SearchParameters', () => {
+      const searchParams = {
+        q: 'query',
+        'f-color': 'red',
+        'f-shape': 'square',
+      };
+
+      const result = toSearchParameters(searchParams);
+
+      expect(result).toEqual({
+        q: 'query',
+        f: {color: ['red'], shape: ['square']},
+      });
+    });
+  });
+
+  describe('when the search parameter value is repeated', () => {
+    it('should convert URLSearchParams to SearchParameters', () => {
+      const urlSearchParams = new URLSearchParams();
+      urlSearchParams.append('q', 'query');
+      urlSearchParams.append('f-color', 'red');
+      urlSearchParams.append('f-color', 'green');
+      urlSearchParams.append('f-shape', 'square');
+
+      const result = toSearchParameters(urlSearchParams);
+
+      expect(result).toEqual({
+        q: 'query',
+        f: {color: ['red', 'green'], shape: ['square']},
+      });
+    });
+
+    it('should convert Record<string, SearchParamValue> to SearchParameters', () => {
+      const searchParams = {
+        q: 'query',
+        'f-color': ['red', 'green'],
+        'f-shape': 'square',
+      };
+
+      const result = toSearchParameters(searchParams);
+
+      expect(result).toEqual({
+        q: 'query',
+        f: {color: ['red', 'green'], shape: ['square']},
+      });
+    });
+  });
+});
+
+describe('serialize', () => {
+  const {serialize} = buildSSRSearchParameterSerializer();
+
+  it('should serialize search parameters to URL', () => {
+    const searchParameters: SearchParameters = {
+      q: 'query',
+      f: {color: ['red', 'green'], shape: ['square']},
+    };
+    const initialUrl = new URL('https://example.com');
+
+    const result = serialize(searchParameters, initialUrl);
+
+    expect(result).toBe(
+      'https://example.com/?q=query&f-color=red&f-color=green&f-shape=square'
+    );
+  });
+
+  it('should not alter url host and route', () => {
+    const searchParameters: SearchParameters = {
+      q: 'query',
+      f: {color: ['red', 'green'], shape: ['square']},
+    };
+    const initialUrl = new URL('https://example.com/custom/route');
+
+    const result = serialize(searchParameters, initialUrl);
+
+    expect(result).toBe(
+      'https://example.com/custom/route?q=query&f-color=red&f-color=green&f-shape=square'
+    );
+  });
+
+  it('should not overwrite existing (non-coveo) search parameters in URL', () => {
```

Comment:
This makes me think: we should ensure that the list of "reserved" Coveo query parameters is documented somewhere so that implementers can take action on their end if needed to avoid conflicts.